tmva/tmva/inc/TMVA/RSofieReader.hxx

```diff
@@ -0,0 +1,243 @@
+/// model operators.
+
+class RSofieReader  {
+
+
+public:
+   /// Create TMVA model from ONNX file
+   /// print level can be 0 (minimal) 1 with info , 2 with all ONNX parsing info
+    RSofieReader(const std::string &path, int verbose = 0)
+   {
+
+      enum EModelType {kONNX, kKeras, kPt, kROOT, kNotDef}; // type of model
+      EModelType type = kNotDef;
+
+      auto pos1 = path.rfind("/");
+      auto pos2 = path.find(".onnx");
+      if (pos2 != std::string::npos) {
+         type = kONNX;
+      } else {
+         pos2 = path.find(".h5");
+         if (pos2 != std::string::npos) {
+             type = kKeras;
+         } else {
+            pos2 = path.find(".pt");
+            if (pos2 != std::string::npos) {
+               type = kPt;
+            }
+            else {
+               pos2 = path.find(".root");
+               if (pos2 != std::string::npos) {
+                  type = kROOT;
+               }
+            }
+         }
+      }
+      if (type == kNotDef) {
+         Error("RSofieReader","Input file is not an ONNX or Keras or PyTorch file");
+         return;
+      }
+      if (pos1 == std::string::npos)
+         pos1 = 0;
+      else
+         pos1 += 1;
+      std::string modelName = path.substr(pos1,pos2-pos1);
+      std::string fileType = path.substr(pos2+1, path.length()-pos2-1);
+      if (verbose) std::cout << "Parsing SOFIE model " << modelName << " of type " << fileType << std::endl;
+
+
+      std::string parserCode;
+      if (type == kONNX) {
+         // check first if we can load the SOFIE parser library
+         if (gSystem->Load("libROOTTMVASofieParser") < 0) {
+            Error("RSofieReader","Cannot use SOFIE with ONNX since libROOTTMVASofieParser is missing");
+            return;
+         }
+         gInterpreter->Declare("#include \"TMVA/RModelParser_ONNX.hxx\"");
+         parserCode += "TMVA::Experimental::SOFIE::RModelParser_ONNX parser; \n";
+         if (verbose == 2)
+            parserCode += "TMVA::Experimental::SOFIE::RModel model = parser.Parse(\"" + path + "\",true); \n";
+         else
+            parserCode += "TMVA::Experimental::SOFIE::RModel model = parser.Parse(\"" + path + "\"); \n";
+      }
+      else if (type == kKeras) {
+         // use Keras direct parser
+         if (gSystem->Load("libPyMVA") < 0) {
+            Error("RSofieReader","Cannot use SOFIE with Keras since libPyMVA is missing");
+            return;
+         }
+         parserCode += "TMVA::Experimental::SOFIE::RModel model = SOFIE::PyKeras::Parse(\"" + path + "\"); \n";
+      }
+      else if (type == kPt) {
+         // use PyTorch direct parser
+         if (gSystem->Load("libPyMVA") < 0) {
+            Error("RSofieReader","Cannot use SOFIE with PyTorch since libPyMVA is missing");
+            return;
+         }
+         parserCode += "TMVA::Experimental::SOFIE::RModel model = SOFIE::PyTorch::Parse(\"" + path + "\"); \n";
```

Comment:
This might not work as expected. The PyTorch parser for SOFIE requires the shapes of the input tensors.